epochStart/shardchain/trigger.go

```diff
@@ -181,14 +190,69 @@ func NewEpochStartTrigger(args *ArgsShardEpochStartTrigger) (*trigger, error) {
 		appStatusHandler:            &statusHandler.NilStatusHandler{},
 	}
 
-	newTrigger.headersPool.RegisterHandler(newTrigger.receivedMetaBlock)
+	t.headersPool.RegisterHandler(t.receivedMetaBlock)
 
-	err := newTrigger.saveState(newTrigger.triggerStateKey)
+	err := t.saveState(t.triggerStateKey)
 	if err != nil {
 		return nil, err
 	}
 
-	return newTrigger, nil
+	t.mapMissingMiniblocks = make(map[string]uint32)
+	go t.requestMissingMiniblocks()
+
+	return t, nil
+}
+
+func (t *trigger) clearMissingMiniblocksMap(epoch uint32) {
+	t.mutMissingMiniblocks.Lock()
+	defer t.mutMissingMiniblocks.Unlock()
+
+	for hash, epochOfMissingMb := range t.mapMissingMiniblocks {
+		if epochOfMissingMb <= epoch {
+			delete(t.mapMissingMiniblocks, hash)
+		}
+	}
+}
+
+func (t *trigger) requestMissingMiniblocks() {
+	for {
+		time.Sleep(sleepTime)
+
+		t.mutMissingMiniblocks.RLock()
+		if len(t.mapMissingMiniblocks) == 0 {
+			t.mutMissingMiniblocks.RUnlock()
+			continue
+		}
+
+		missingMiniblocks := make([][]byte, len(t.mapMissingMiniblocks))
+		for hash := range t.mapMissingMiniblocks {
+			missingMiniblocks = append(missingMiniblocks, []byte(hash))
+			log.Debug("trigger.requestMissingMiniblocks", "hash", []byte(hash))
+		}
+		t.mutMissingMiniblocks.RUnlock()
+
+		go t.requestHandler.RequestMiniBlocks(core.MetachainShardId, missingMiniblocks)
+		time.Sleep(waitTime)
+		t.updateMissingMiniblocks()
+	}
+}
+
+func (t *trigger) updateMissingMiniblocks() {
+	t.mutMissingMiniblocks.Lock()
+	for hash := range t.mapMissingMiniblocks {
```

Comment:
this for is not needed if the delete from mapMissingMiniBLocks is done at receive time - this is time consuming.

Reply:
if you make a register handler to the received miniblocks then this for is not needed - only if numMissingMiniBlocks == 0

Reply:
You can not do this there, as for each miniblock received you would call updateTrigger as numMissingMiniBlocks would be 0 fro most of time